tensorboard/uploader/exporter_test.py

```diff
@@ -227,6 +248,214 @@ def outdir_files():
+        summary_metadata = summary_pb2.SummaryMetadata.FromString(
+            base64.b64decode(datum.pop("summary_metadata"))
+        )
+        expected_summary_metadata = summary_pb2.SummaryMetadata(
+            data_class=summary_pb2.DATA_CLASS_BLOB_SEQUENCE
+        )
+        self.assertEqual(summary_metadata, expected_summary_metadata)
+        points = datum.pop("points")
+        self.assertEqual(points["steps"], [0])
+        self.assertEqual(points["wall_times"], [1571084520.862939144])
+        # The 1st blob is finished; the 2nd is unfinished.
+        self.assertEqual(
+            points["blob_file_paths"], [["blobs/blob_train_blob.bin", None]]
+        )
+
+        datum = jsons[1]
+        self.assertEqual(datum.pop("run"), "test")
+        self.assertEqual(datum.pop("tag"), "__default_graph__")
+        summary_metadata = summary_pb2.SummaryMetadata.FromString(
+            base64.b64decode(datum.pop("summary_metadata"))
+        )
+        self.assertEqual(summary_metadata, expected_summary_metadata)
+        points = datum.pop("points")
+        self.assertEqual(points["steps"], [0])
+        self.assertEqual(points["wall_times"], [1571084520.862939144])
+        # `None` blob file path indicates an unfinished blob.
+        self.assertEqual(points["blob_file_paths"], [[None]])
+
+        # Check the BLOB files.
+        with open(
+            os.path.join(
+                outdir, "experiment_123", "blobs", "blob_train_blob.bin"
+            ),
+            "rb",
+        ) as f:
+            self.assertEqual(f.read(), b"43218765")
+
+        # Check call to StreamBlobData.
+        expected_blob_data_request = export_service_pb2.StreamBlobDataRequest(
+            blob_id="train_blob"
+        )
+        mock_api_client.StreamBlobData.assert_called_once_with(
+            expected_blob_data_request, metadata=grpc_util.version_metadata()
+        )
+
+        # Test the case where blob streaming errors out.
+        self.assertEqual(next(generator), "456")
+        # Check the blob_sequences.json file.
+        with open(
+            os.path.join(outdir, "experiment_456", "blob_sequences.json")
+        ) as infile:
+            jsons = [json.loads(line) for line in infile]
+        self.assertLen(jsons, 2)
+
+        datum = jsons[0]
+        self.assertEqual(datum.pop("run"), "train")
+        self.assertEqual(datum.pop("tag"), "__default_graph__")
+        summary_metadata = summary_pb2.SummaryMetadata.FromString(
+            base64.b64decode(datum.pop("summary_metadata"))
+        )
+        self.assertEqual(summary_metadata, expected_summary_metadata)
+        points = datum.pop("points")
+        self.assertEqual(points["steps"], [0])
+        self.assertEqual(points["wall_times"], [1571084520.862939144])
+        # `None` represents the blob that experienced error during downloading
+        # and hence is missing.
+        self.assertEqual(points["blob_file_paths"], [[None, None]])
```

Comment:
In the other tests we check `self.assertEqual(datum, {})` after reading
each field to verify that we didn’t get anything unexpected; did we mean
to do the same thing here and below? (Otherwise, I don’t see the reason
for using `pop` instead of just subscripting.)

Reply:
Good point. I revised the new test code in this PR to do `pop()` and empty-dict assertion.